### PR TITLE
[Driver] Support -mcmodel= for LoongArch

### PR DESCRIPTION
**Pull request by SixWeining**
7e42545 rejects unsupported mcmodel options, but normal/medium/extreme should be supported models for LoongArch according to [gcc document](https://gcc.gnu.org/onlinedocs/gcc/LoongArch-Options.html).

The mappings among `gcc`, `clang driver`, `clang cc1` and `LLVM (i.e. llc --code-model=)` are:

|     gcc      |  clang driver |  clang cc1   |    LLVM    |
| ---------  | -------------- |------------- |----------- |
| normal    |      normal     |   small        |     small    |
| medium  |     medium    |   medium   |    medium |
| extreme  |     extreme     |  large         |     large     |

Link: https://reviews.llvm.org/D150522

### CODE REVIEW

**Comment by SixWeining** [comment]
cc @xen0n @xry111

**Comment by xry111** [comment]
> And AFAIK, gcc side doesn't plan to implement the "large" code model.

Why did we distinguish "large" and "extreme" in the first place?  If we don't need a different "large" code model then I guess we should make it an alias of "extreme" for GCC too.

**Comment by SixWeining** [comment]
> > And AFAIK, gcc side doesn't plan to implement the "large" code model.
> 
> Why did we distinguish "large" and "extreme" in the first place? If we don't need a different "large" code model then I guess we should make it an alias of "extreme" for GCC too.

@ChenghuaXu @chenglulu326 What do you think?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -5739,6 +5739,16 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
       if (CM == "large" && RelocationModel != llvm::Reloc::Static)
         D.Diag(diag::err_drv_argument_only_allowed_with)
             << A->getAsString(Args) << "-fno-pic";
+    } else if (Triple.isLoongArch()) {
+      CM = llvm::StringSwitch<StringRef>(CM)
+               .Case("normal", "small")
+               .Case("extreme", "large")
+               .Default(CM);
+      if (CM == "large" &&
+          Args.hasFlag(options::OPT_fplt, options::OPT_fno_plt, false))
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/Clang.cpp:5748]: I consider -fno-pic/-fpic a more fundamental difference than -mcmodel=, so I place the aarch64 compatibility check in this large `if` condition.

-mcmodel= is a more fundamental difference than -fplt, so the -fplt compatibility check should probably belong to -fplt. You can move 

```
if (Args.hasFlag(options::OPT_fno_plt, options::OPT_fplt, false)) {
```
below CodeModel check and add the diagnostic there.

Comment (MaskRay) [reply clang/lib/Driver/ToolChains/Clang.cpp:5748]: nvm. this is fine

Comment (MaskRay) [inline clang/lib/Driver/ToolChains/Clang.cpp:5748]: `hasFlagNoClaim` is slightly better

**Comment by MaskRay** [comment]
[gcc.gnu.org/onlinedocs/gcc/LoongArch-Options.html](https://gcc.gnu.org/onlinedocs/gcc/LoongArch-Options.html) says

‘large (Not implemented yet)’
‘extreme’
This mode does not limit the size of the code segment and data segment. The -mcmodel=extreme option is incompatible with -fplt, and it requires -mexplicit-relocs=always.

Is there a psABI defining 'extreme' or 'large'?

File: clang/test/Driver/mcmodel.c [hunk 2]
```diff
@@ -15,6 +15,15 @@
 // RUN: not %clang -### -c --target=aarch64 -mcmodel=medium %s 2>&1 | FileCheck --check-prefix=ERR-MEDIUM %s
 // RUN: not %clang -### -c --target=aarch64 -mcmodel=kernel %s 2>&1 | FileCheck --check-prefix=ERR-KERNEL %s
 // RUN: not %clang --target=aarch64_32-linux -### -S -mcmodel=small %s 2>&1 | FileCheck --check-prefix=ERR-AARCH64_32 %s
+// RUN: %clang --target=loongarch64 -### -S -mcmodel=small %s 2>&1 | FileCheck --check-prefix=SMALL %s
+// RUN: %clang --target=loongarch64 -### -S -mcmodel=normal %s 2>&1 | FileCheck --check-prefix=SMALL %s
```
Comment (MaskRay) [inline clang/test/Driver/mcmodel.c:19]: If `-mcmodel=normal` is canonical, we probably should not add Clang-specific alias `small`. Reject `normal`.

**Comment by SixWeining** [comment]
> [gcc.gnu.org/onlinedocs/gcc/LoongArch-Options.html](https://gcc.gnu.org/onlinedocs/gcc/LoongArch-Options.html) says
> 
> ‘large (Not implemented yet)’ ‘extreme’ This mode does not limit the size of the code segment and data segment. The -mcmodel=extreme option is incompatible with -fplt, and it requires -mexplicit-relocs=always.
> 
> Is there a psABI defining 'extreme' or 'large'?

Currently there is no definition in psABI except in gcc doc: https://gcc.gnu.org/onlinedocs/gcc/LoongArch-Options.html. Maybe we can define them in next psABI version.

**Comment by zjiaz** [comment]
FWIW, `mcmodel=medium` option is used in [chromium's build config file](https://source.chromium.org/chromium/chromium/src/+/main:build/config/compiler/BUILD.gn;l=1451?q=mcmodel&ss=chromium%2Fchromium%2Fsrc:build%2F), and now chromium's llvm toolchain has been affected by this issue:
> clang++: error: unsupported argument 'medium' to option '-mcmodel=' for target 'loongarch64-unknown-linux-gnu'

**Review by wangleiat — APPROVED** [review]
LGTM

**Pull request merged**